### PR TITLE
cmd/gomobile: support Catalyst and Apple Silicon

### PR DESCRIPTION
**Pull request by waylybaye**
This PR is derived from golang/mobile#45 .

I changed the build process and it will create 3 frameworks for ios, simulator and catalyst. Every framework has a fat library contains both arm64 and x86_64 (except ios that only needs arm64). 

Finally create one xcframework from all 3 fat frameworks and this xcframework contains every platform and architecture.

### CODE REVIEW

**Comment by waylybaye** [comment]
@googlebot I signed it!

**Comment by aeddi** [comment]
> All (the pull request submitter and all commit authors) CLAs are signed, **but** one or more commits were authored or co-authored by someone other than the pull request submitter.
> 
> We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that by leaving a comment that contains only `@googlebot I consent.` in this pull request.
> 
> _Note to project maintainer:_ There may be cases where the author cannot leave a comment, or the comment is not properly detected as consent. In those cases, you can manually confirm consent of the commit author(s), and set the `cla` label to `yes` (if enabled on your project).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fgolang%2Fmobile%2Fpull%2F63) for more info**.

@dpwiese Can you confirm your consent with the bot? :)

**Comment by dpwiese** [comment]
@googlebot I consent.

**Review by dpwiese — COMMENTED** [review]
I'm happy to see this moving along, and think considering targets and architectures separately (as opposed to just treating `catalyst` as a new architecture as I had) is a much more elegant approach! And great to see Apple Silicon support added. 🎉 

I've built my `Sample.xcframework` from this branch it it works on the target/architecture combos I've tested it quickly on: my old `x86_64` mac with both simulator and catalyst, and my `arm64` iPhone. Seems to work as expected. 👍 

As far as tests, `TestBindIOS` fails now, but that's expected and should be a quick fix of updating the templates. I also have the `TestBindWithGoModules` tests failing. And finally `TestIOSBuild` is failing too, but that was already the case as mentioned in https://github.com/golang/mobile/pull/45#issue-397087305. I'm not sure how much effort should be placed towards these tests given they are failing already, but it would be very nice to get this helpful PR merged in. 🙂

File: cmd/gomobile/env.go [hunk 1]
```diff
@@ -28,14 +28,41 @@ var (
 func allArchs(targetOS string) []string {
 	switch targetOS {
 	case "ios":
-		return []string{"arm64", "amd64"}
+		return []string{"arm64", "amd64", "catalyst"}
```
[the remark this answers is not in the document]
Comment (waylybaye) [reply cmd/gomobile/env.go]: Yes, I've removed it 😄

**Comment by michaellady** [comment]
Would this PR make .xcframeworks built by gomobile compatible with Swift Package Manager (SPM)?

**Comment by ydnar** [comment]
I’m trying to use this fork with a `replace` directive in my project’s `go.mod` file, and it fails with an error:

```
go: cloud.google.com/go@v0.37.0: missing go.sum entry; to add it:
	go mod download cloud.google.com/go
```

The full command output:

```
❯ gomobile bind -target ios ./mobile/ping
[debug] goenv:
[GOOS=darwin GOARCH=arm64 CC=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang CXX=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ CGO_CFLAGS=-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.4.sdk -mios-simulator-version-min=7.0 -fembed-bitcode -arch arm64 CGO_CXXFLAGS=-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.4.sdk -mios-simulator-version-min=7.0 -fembed-bitcode -arch arm64 CGO_LDFLAGS=-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.4.sdk -mios-simulator-version-min=7.0 -fembed-bitcode -arch arm64 CGO_ENABLED=1 ARCH=arm64 GOPATH=/var/folders/6g/1gng3zts0t39s_qbtt7p0wsc0000gn/T/gomobile-work-204926828:/Users/ydnar/go]
gomobile: darwin-arm64: go build -tags ios -buildmode=c-archive -o /var/folders/6g/1gng3zts0t39s_qbtt7p0wsc0000gn/T/gomobile-work-204926828/ping-arm64.a ./gobind failed: exit status 1
go: cloud.google.com/go@v0.37.0: missing go.sum entry; to add it:
	go mod download cloud.google.com/go
```

Go 1.16 on macOS 11.3 beta, using Xcode 12.4 on an Intel MB Pro.

**Comment by nadimkobeissi** [comment]
This is a pretty exciting and important PR, any idea when it will get merged?

Currently, if I try to install it manually, I get this strange error when I run `gomobile bind`:

```
[debug] goenv:
[GOOS=darwin GOARCH=arm64 CC=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang CXX=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ CGO_CFLAGS=-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.5.sdk -mios-simulator-version-min=7.0 -fembed-bitcode -arch arm64 CGO_CXXFLAGS=-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.5.sdk -mios-simulator-version-min=7.0 -fembed-bitcode -arch arm64 CGO_LDFLAGS=-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.5.sdk -mios-simulator-version-min=7.0 -fembed-bitcode -arch arm64 CGO_ENABLED=1 ARCH=arm64 GOPATH=/var/folders/sk/85yp792d7xb6nlmnzn23_jf40000gn/T/gomobile-work-735475716:/Users/nadim/go]
gomobile: darwin-arm64: go build -tags ios -gcflags -e -ldflags -s -w -trimpath -buildmode=c-archive -o /var/folders/sk/85yp792d7xb6nlmnzn23_jf40000gn/T/gomobile-work-735475716/MyGoLibrary-arm64.a ./gobind failed: exit status 1
go: github.com/BurntSushi/xgb@v0.0.0-20210121224620-deaf085860bc: missing go.sum entry; to add it:
	go mod download github.com/BurntSushi/xgb
```

Error aside, it would be really cool to see this merged and pushed into stable.

**Comment by ydnar** [comment]
> This is a pretty exciting and important PR, any idea when it will get merged?
> 
> 
> 
> Currently, if I try to install it manually, I get this strange error when I run `gomobile bind`:
> 
> 
> 
> ```
> 
> [debug] goenv:
> 
> [GOOS=darwin GOARCH=arm64 CC=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang CXX=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ CGO_CFLAGS=-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.5.sdk -mios-simulator-version-min=7.0 -fembed-bitcode -arch arm64 CGO_CXXFLAGS=-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.5.sdk -mios-simulator-version-min=7.0 -fembed-bitcode -arch arm64 CGO_LDFLAGS=-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.5.sdk -mios-simulator-version-min=7.0 -fembed-bitcode -arch arm64 CGO_ENABLED=1 ARCH=arm64 GOPATH=/var/folders/sk/85yp792d7xb6nlmnzn23_jf40000gn/T/gomobile-work-735475716:/Users/nadim/go]
> 
> gomobile: darwin-arm64: go build -tags ios -gcflags -e -ldflags -s -w -trimpath -buildmode=c-archive -o /var/folders/sk/85yp792d7xb6nlmnzn23_jf40000gn/T/gomobile-work-735475716/MyGoLibrary-arm64.a ./gobind failed: exit status 1
> 
> go: github.com/BurntSushi/xgb@v0.0.0-20210121224620-deaf085860bc: missing go.sum entry; to add it:
> 
> 	go mod download github.com/BurntSushi/xgb
> 
> ```
> 
> 
> 
> Error aside, it would be really cool to see this merged and pushed into stable.

See #65 for a build that fixes that issue

**Comment by gadireddi226** [comment]
When this PR will be merged ?

**Comment by deepumukundan** [comment]
Guess we can close this PR now that the [fork](https://github.com/golang/mobile/pull/65) with these and additional changes has been merged in! Finally closure!

**Comment by waylybaye** [comment]
closed since #65 has been merged

**Comment by nadimkobeissi** [comment]
Thank you so much!

**Pull request closed without merging**